src/common/checksum.c

```diff
@@ -20,218 +20,509 @@
+	case DAOS_PROP_CO_CSUM_CRC16:
+		return CSUM_TYPE_ISAL_CRC16_T10DIF;
+	case DAOS_PROP_CO_CSUM_CRC32:
+		return CSUM_TYPE_ISAL_CRC32_ISCSI;
+	case DAOS_PROP_CO_CSUM_CRC64:
+		return CSUM_TYPE_ISAL_CRC64_REFL;
+	case DAOS_PROP_CO_CSUM_SHA1:
+		return CSUM_TYPE_ISAL_SHA1;
 
-	rc = mchecksum_init(cs_name, &cs_obj->dc_csum);
-	if (rc < 0) {
-		D_ERROR("Error in initializing checksum\n");
-		return -DER_NOMEM;
+	default:
+		return CSUM_TYPE_UNKNOWN;
 	}
-#endif
-	cs_obj->dc_init = 1;
-	memset(cs_obj->dc_buf, 0, DAOS_CSUM_SIZE);
-	D_DEBUG(DB_IO, "Initialize checksum=%s\n", dict->cs_name);
-	return 0;
 }
 
-inline int
-daos_csum_reset(daos_csum_t *cs_obj)
-{
-	if (!cs_obj->dc_init)
-		return -DER_UNINIT;
-#if defined(__x86_64__)
-	memset(cs_obj->dc_buf, 0, DAOS_CSUM_SIZE);
-#else
-	int  rc;
+/**
+ * ---------------------------------------------------------------------------
+ * Algorithms
+ * ---------------------------------------------------------------------------
+ */
 
-	rc = mchecksum_reset(cs_obj->dc_csum);
-	if (rc < 0) {
-		D_ERROR("Error resetting mchecksum: %d\n", rc);
-		return -DER_UNINIT;
-	}
-#endif
+/** CSUM_TYPE_ISAL_CRC16_T10DIF*/
+static int
+crc16_update(struct daos_csummer *obj, uint8_t *buf, size_t buf_len)
+{
+	uint16_t *crc16 = (uint16_t *) obj->dcs_csum_buf;
+	*crc16 = crc16_t10dif(*crc16, buf, (int) buf_len);
 	return 0;
 }
 
-inline int
-daos_csum_free(daos_csum_t *cs_obj)
+struct csum_ft crc16_algo = {
+	.update = crc16_update,
+	.csum_len = sizeof(uint16_t),
+	.name = "crc16"
+};
+
+/** CSUM_TYPE_ISAL_CRC32_ISCSI */
+static int
+crc32_update(struct daos_csummer *obj, uint8_t *buf, size_t buf_len)
 {
-#if !defined(__x86_64__)
-	mchecksum_destroy(cs_obj->dc_csum);
-#endif
+	uint32_t *crc32 = (uint32_t *) obj->dcs_csum_buf;
+	*crc32 = crc32_iscsi(buf, (int) buf_len, *crc32);
 	return 0;
 }
 
-inline daos_size_t
-daos_csum_get_size(const daos_csum_t *csum)
+struct csum_ft crc32_algo = {
+		.update = crc32_update,
+		.csum_len = sizeof(uint32_t),
+		.name = "crc32"
+	};
+
+/** CSUM_TYPE_ISAL_CRC64_REFL */
+static int
+crc64_update(struct daos_csummer *obj,
+		 uint8_t *buf, size_t buf_len)
 {
-#if defined(__x86_64__)
-	return csum_dict[csum->dc_csum].cs_size;
-#else
-	return mchecksum_get_size(csum->dc_csum);
-#endif
+	uint64_t *csum = (uint64_t *)obj->dcs_csum_buf;
+
+	*csum = crc64_ecma_refl(*csum, buf, buf_len);
+	return 0;
 }
 
-inline int
-daos_csum_get(daos_csum_t *csum, daos_csum_buf_t *csum_buf)
+struct csum_ft crc64_algo = {
+		.update = crc64_update,
+		.csum_len = sizeof(uint64_t),
+		.name = "crc64"
+	};
+
+/** CSUM_TYPE_ISAL_SHA1 */
+static int
+sha1_init(struct daos_csummer *obj)
 {
+	struct mh_sha1_ctx *ctx;
 
-#if defined(__x86_64__)
-	if (csum_buf->cs_buf_len != csum_dict[csum->dc_csum].cs_size) {
-		D_ERROR("Incorrect result buffer size provided\n");
-		return -DER_INVAL;
-	}
-	memcpy(csum_buf->cs_csum, csum->dc_buf,
-	       csum_dict[csum->dc_csum].cs_size);
+	D_ALLOC(ctx, sizeof(struct mh_sha1_ctx));
```

Comment:
need a NULL check here and a -DER_NOMEM.   Error message will be taken care of by D_ALLOC

Reply:
fixed